### ISSUE TITLE
Go through all BTC CVE

### ISSUE THREAD

**Issue by noescape00** [stratisproject/StratisBitcoinFullNode#1822]

**Comment by bokobza** [comment]
This task can be divided into 2 parts:
1. Create a list of known bitcoin vulnerabilities
2. check whether the source code has protection mechanism against each and every one of those vulnerabilities

Also, for each vulnerability, add comments supporting your findings.

**Comment by Aprogiena** [comment]
![image](https://user-images.githubusercontent.com/30052221/43717951-0ce0ce30-9981-11e8-86e3-0c87c3f27af0.png)

**Comment by bokobza** [comment]
CVE-2012-4684 We don't have alerts anymore so that's easy ✔️

**Comment by DanGould** [comment]
@Aprogiena Where is that list coming from? [bitcoin.it wiki's lost](https://en.bitcoin.it/wiki/Common_Vulnerabilities_and_Exposures) seems comprehensive

**Comment by Aprogiena** [comment]
https://youtu.be/Cpid31c6HZc?t=2h52m33s

**Comment by DanGould** [comment]
Progress thus far noted: Should be through most of the bitcoin list by the weekend. Still must check in on the Etherium CVEs. I will update this comment with the status of the CVEs
| CVE                                                                                            | Affects                          | Flaw                                                  | Comments                                                                                           |
|------------------------------------------------------------------------------------------------|----------------------------------|-------------------------------------------------------|----------------------------------------------------------------------------------------------------|
| Pre-BIP protocol changes                                                                       | all clients                      | various softforks & hardforks                         | Do we validate the Protocol?                                                                       |
| [CVE 2010 5137](https://en.bitcoin.it/wiki/Common_Vulnerabilities_and_Exposures#CVE-2010-5137) |  wxBitcoin & bitcoind            | OP_LSHIFT crash                                       | Disabled in `ScriptEvaluationContext.cs:696`                                                       |
| [CVE-2010-5141](https://en.bitcoin.it/wiki/Common_Vulnerabilities_and_Exposures#CVE-2010-5141) | wxBitcoin & bitcoind             | OP_RETURN can steal funds                             |                                                                                                    |
| [CVE-2010-5138](https://en.bitcoin.it/wiki/Common_Vulnerabilities_and_Exposures#CVE-2010-5138) |                                  | unlimited SigOp Dos                                   | `Check SigOpsRuleTest.cs`                                                                          |
| [CVE-2010-5139](https://en.bitcoin.it/wiki/Value_overflow_incident)                            |                                  | Combined output overflow                              | Check for negative or overflow output values in `Transaction.cs`                                   |
| [CVE-2010-5140](https://en.bitcoin.it/wiki/Common_Vulnerabilities_and_Exposures#CVE-2010-5140) | wxBitcoin & bitcoind             | never confirming txs                                  |  No micropayment fee `TransactionBuilderContext.MinConfirmations` defaults to 1                    |
| [CVE-2011-4447](https://en.bitcoin.it/wiki/Common_Vulnerabilities_and_Exposures#CVE-2011-4447) | wxBitcoin & bitcoind             | Wallet non-encryption                                 |                                                                                                    |
| [CVE-2012-1909](https://en.bitcoin.it/wiki/Common_Vulnerabilities_and_Exposures#CVE-2012-1909) | Bitcoin protocol and all clients | Transaction overwriting                               |  `DeploymentFlags.EnforceBip30` Checks block height & hash for 2 bitcoin blockchain exceptions |
| [CVE-2012-1910](https://en.bitcoin.it/wiki/Common_Vulnerabilities_and_Exposures#CVE-2012-1910) | bitcoinqt & bitcoind for windows | Fail to use MinGW multithread safe exception handling | StratisBitcoinFullNode doesn't use MinGW                                                           |
| [BIP-0016](https://en.bitcoin.it/wiki/Common_Vulnerabilities_and_Exposures#BIP-0016)           | All bitcoin clients              | Mandatory P2SH                                        | Thanks NBitcoin and a mature existing blockchain                                                   |
| [CVE-2012-2459](https://en.bitcoin.it/wiki/Common_Vulnerabilities_and_Exposures#CVE-2012-2459) | bitcoind & bitcoin-Qt            | block hash collisions via merkle root                 | `BlockMerkleRootRule.cs`                                                                           |
| [CVE-2012-3789](https://en.bitcoin.it/wiki/CVE-2012-3789)                                      | bitcoind & bitcoin-Qt            | Lack of orphan Txn resource limits                    |                                                                                                    |

**Comment by bokobza** [comment]
![image](https://user-images.githubusercontent.com/1867877/44205778-790a3300-a14f-11e8-9548-cb65bdee3b37.png)

The wallet encrypts the private key of the HD wallet using `Key.GetEncryptedBitcoinSecret(password, network)` in NBitcoin who in turn uses `SCrypt.BitcoinComputeDerivedKey` which is an implementation of BIP38.

**Comment by dangershony** [comment]
Great work @DanGould we will verify all the findings when the node is stable.

**Comment by DanGould** [comment]
After speaking with Ethan, I found out that the Core devs' policy is now to **not** disclose CVEs because they don't want to be responsible for alts' follies. Notice reporting drops significantly after 2015.

He mentioned their new practice to find some small bug which, when fixed, patches a vulnerability without ever disclosing that vulnerability. He mentioned a few in particular with no documentation and that core devs involved with MIT DCI speak openly about this practice as evidence. This practice is reflected in the little information on many bugs provided on the bitcoin.it wiki 

### tldr;
The most historically vulnerable attack surfaces are the Mempool and Transaction verification. I advise we compare present /bitcoin/bitcoin's implementations of these functions with SBFN's and closely match them to avoid exposure.

**Comment by Fishbiscuit** [comment]
May I ask if there is a list of Bitcoin versions that resolve these CVEs?

**Comment by DanGould** [comment]
> May I ask if there is a list of Bitcoin versions that resolve these CVEs?

https://en.bitcoin.it/wiki/Common_Vulnerabilities_and_Exposures Click the "net" link on the right side of the row corresponding to the CVE. Shows the list of versions vulnerable at the bottom. Generally, if you search the CVE on GitHub you find the issue which addressed the vulnerability

**Comment by EthanHeilman** [comment]
**Disclaimer:** I in no way speak for the core devs

I just want to poke my head in to say that I am not aware of any official policy of the core devs to **not** disclose vulnerabilities. However there was a period in which a number of vulnerabilities were patched but not disclosed. This was likely because of a concern of them being exploited against unpatched nodes in the wild. This has not been the case for a while and last few vulnerabilities have been disclosed and assigned CVE numbers for instance [CVE-2018-17144](https://bitcoincore.org/en/2018/09/20/notice/).

The closest thing to a policy is the [Security_md](https://github.com/bitcoin/bitcoin/blob/master/SECURITY.md) in the bitcoin git repo and it does not take any stance on disclosure.
